python/triton/compiler/compiler.py

```diff
@@ -142,9 +143,18 @@ def parse_options(self):
         return dict()
 
 
+def iter_modules_recursively(paths):
+    for path in paths:
+        for lib in pkgutil.iter_modules([path]):
```

Comment:
Is this not https://docs.python.org/3/library/pkgutil.html#pkgutil.walk_packages ?

Reply:
Oh I didn't know that. Looks like it's what we need, but when I tried it, it also gave a lot other dependencies like

python3.11/site-packages/nvidia/nvjitlink/lib/__init__.py'>
python3.11/site-packages/nvidia/nvtx/__init__.py'>
python3.11/site-packages/nvidia/nvtx/include/__init__.py'>
python3.11/site-packages/nvidia/nvtx/lib/__init__.py'

Reply:
Can you consult with the other folks at Meta and figure out what's the right way to do this?  I'm no expert.

Reply:
Does it work for you? Just want to see if the API is supposed to work, or it's an issue with our server environment. We are using pip packages internally, not sure how specific it is. Thanks!

Reply:
Sorry, I don't have cycles for a lot of back-and-forth on the bug right now.

Reply:
OK, I switched to a different server and the problem still exist. Basically the following code 

```
 for lib in pkgutil.walk_packages([tp_backends_path]):
        with open(lib.module_finder.find_spec(lib.name).origin, "rb") as f:
            print(f)
            contents += [hashlib.sha256(f.read()).hexdigest()]
```

yielded something like:

```
<_io.BufferedReader name='triton/python/triton/backends/compiler.py'>
<_io.BufferedReader name='triton/python/triton/backends/driver.py'>
<_io.BufferedReader name='triton/python/triton/backends/nvidia/__init__.py'>
<_io.BufferedReader name='python3.11/site-packages/nvidia/cublas/__init__.py'>
<_io.BufferedReader name='python3.11/site-packages/nvidia/cublas/include/__init__.py'>
```

I can get rid of the last two unrelated packages by providing the `prefix` argument to `walk_packages`

```
 for lib in pkgutil.walk_packages([tp_backends_path], 'backends.'):
        with open(lib.module_finder.find_spec(lib.name).origin, "rb") as f:
            print(f)
            contents += [hashlib.sha256(f.read()).hexdigest()]
```

, but the following real python files instead of the `backends/nvidia/__init__.py` are still missing, which can be recovered by the current patch.

```
<_io.BufferedReader name='triton/python/triton/backends/nvidia/compiler.py'>
<_io.BufferedReader name='triton/python/triton/backends/nvidia/driver.py'>
```

By briefly search around, there are a few issues and bugs reported to this API, e..g, `https://github.com/python/cpython/issues/92229`, I'm not sure how reliable the API is for now.